### PR TITLE
Root-level message listeners

### PR DESCRIPTION
**Pull request by garyb**

### CODE REVIEW

File: src/Halogen/Aff/Driver.purs [hunk 1]
```diff
@@ -43,12 +44,17 @@ import Halogen.Query.HalogenM (HalogenM(..), HalogenF(..), HalogenAp(..))
...
--- | Type alias for driver functions generated by `runUI` - a driver takes an
--- | input of the query algebra (`f`) and returns an `Aff` that returns when
--- | query has been fulfilled.
-type Driver f eff = f ~> Aff (HalogenEffects eff)
-
-type DSL s f m eff p o = HalogenF s f m p o (Aff (HalogenEffects eff))
+-- | A record produced when the root component in a Halogen UI has been run.
+-- | `query` allows external sources to query the root component and `subscribe`
+-- | allows external consumers to receive messages raised by the root component.
+type HalogenIO f o m =
```
Comment (natefaubion) [inline src/Halogen/Aff/Driver.purs]: This type is not Aff specific. Should it be defined elsewhere?

File: src/Halogen/Aff/Driver.purs [hunk 2]
```diff
@@ -88,28 +94,52 @@ runUI
...
+  subscribe
+    :: Ref Int
+    -> Ref (M.Map Int (AV.AVar (Either o Unit)))
+    -> ({ producer :: CR.Producer o (Aff (HalogenEffects eff)) Unit
+        , unsubscribe :: Aff (HalogenEffects eff) Boolean
+        } -> Aff (HalogenEffects eff) Unit)
+    -> Aff (HalogenEffects eff) Unit
+  subscribe fresh ref k = do
+    listenerId <- liftEff $ readRef fresh
+    liftEff $ modifyRef fresh (_ + 1)
```
Comment (natefaubion) [inline src/Halogen/Aff/Driver.purs]: You could reduce the liftEffs:
```purescript
listenerId <- liftEff (readRef fresh <* modifyRef fresh (plus 1))
```

This is just a matter of taste though :P

File: examples/counter/src/Main.purs [hunk 3]
```diff
@@ -41,8 +41,8 @@ ui = H.component { render, eval, initialState }
 main :: Eff (H.HalogenEffects ()) Unit
 main = runHalogenAff do
   body <- awaitBody
-  driver <- runUI ui body
-  setInterval 1000 $ driver (H.action Tick)
+  app <- runUI ui body
+  setInterval 1000 $ app.query (H.action Tick)
```
Comment (natefaubion) [inline examples/counter/src/Main.purs]: We could probably use an example that listens to events as well.

File: src/Halogen/Aff/Driver.purs (hunk 1 above)
Comment (garyb) [inline src/Halogen/Aff/Driver.purs]: I pondered that too. Maybe just in the main `Halogen` module?

File: examples/counter/src/Main.purs (hunk 3 above)
Comment (garyb) [inline examples/counter/src/Main.purs]: Yeah, I was going to add that separately, will do so while working on the docs.

File: src/Halogen/Aff/Driver.purs (hunk 2 above)
Comment (garyb) [inline src/Halogen/Aff/Driver.purs]: How about
``` purescript
listenerId <- liftEff do
  lid <- readRef fresh
  modifyRef fresh (_ + 1)
  pure lid
```
We get "magic do" optimisations that way? Or I could just leave it 😛

File: src/Halogen/Aff/Driver.purs (hunk 1 above)
Comment (natefaubion) [inline src/Halogen/Aff/Driver.purs]: That won't cause circular dependency issues?

File: src/Halogen/Aff/Driver.purs (hunk 2 above)
Comment (natefaubion) [inline src/Halogen/Aff/Driver.purs]: Not a sticking point :)

File: examples/counter/src/Main.purs (hunk 3 above)
Comment (natefaubion) [inline examples/counter/src/Main.purs]: I just meant to show that it does indeed work, before merging :D

Comment (natefaubion) [inline examples/counter/src/Main.purs]: Like, I don't care if its even to this one. This could be the "IO" example.

Comment (garyb) [inline examples/counter/src/Main.purs]: Ah ok, I did check that! https://gist.github.com/garyb/e6aacc4019a1781d846a2a82e203c41b

File: src/Halogen/Aff/Driver.purs (hunk 1 above)
Comment (garyb) [inline src/Halogen/Aff/Driver.purs]: No, the driver stuff isn't re-exported from `Halogen`, and `HalogenIO`'s types are not driver spectific either.

**Pull request merged**